ls1intum/Artemis
# `Discussion`: Develop messages page for courses

### Checklist
#### General


#### Server
#### Client

### Motivation and Context
Although it is possible to communicate via channels about existing course content (e.g. lectures, exercises) and course-wide context (e.g. TECH_SUPPORT, ORGANIZATION, ...) with the previous course discussions page, the communication is rather expected to be formal and is public to every participant of the course. Furthermore, tutors are able to edit or delete posts of others, which would not be appropriate in a one-to-one discussion.

For this, we introduce a new "messages" page for courses, allowing course participants to send direct messages to each other.



### Description
Extending the previous discussion page, we introduce a closable sidebar, which allows searching and selecting other registered participants of the course (excluding the logged-in user) to communicate. Upon selection, the conversation between the users are fetched and shown in a descending sending time by default, as it is done in most of the messaging platforms. Instead of the pop-up that was used for creating new posts with mandatory fields like title etc., we introduce an inline input with adjustable height. Users are still able to update/delete their own messages and tutors are no longer able to manipulate the posts of others.

The current implementation is still under development, hence the page will only be available in non-production environments. If there is an update on the messages a user has fetched, the updates are immediately reflected to the user via WebSocket subscription. However, there is currently no implementation for notifying users about new messages while they are not displaying the relevant conversation.

![Course Messages UML Class Diagram](https://user-images.githubusercontent.com/67344158/162987041-46ee630a-489e-4c43-98e2-c451c8011f5d.png)

### New Entity Objects
**Conversation**: A new context which posts can be affiliated to. It is required to distinguish different groups of messages and we need to keep it separate from ConversationParticipant because, we want to allow private group chats in future developments and need flexibility for that. Each conversation must consist of a Course and two ConversationParticipants (will be increased in the future to support private group messages).

**ConversationParticipant**: Contains user-specific information of users who participate in a conversation. The ConversationParticipant of users are passed on to other participants of the Conversation with sensitive information filtered out, so they can see with whom they are messaging. ConversationParticipant consists of two variables; lastRead and closed, which will be utilized in future developments.

### Steps for Testing

Prerequisites:
- 1 course with discussions enabled
- 1 Student registered and browsed the course
- 1 Tutor registered and browsed the course


1. Log in to Artemis
2. Navigate to Course
3. Navigate to the Messages Page
4. From the searchbar on the left, search for another participant of the course to message.
5. From the list of users displayed, select one to create a new conversation.
6. Verify that the search bar was cleared after user selection and the created conversation is selected by default.
7. From the inline input, send messages to the other user and verify that they can display them appropriately.
8. Edit & delete messages and verify that they are updated for the recipient as well.
9. Make sure that the tutor can only update and delete their own messages.
10. Navigate to the Discussion Page
11. Verify that all functionality still works correctly.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





ConversationService.java: 75% | 96.07%
post-overview.component.ts: 84.37% | 96.55%
course-messages.component.ts: 100% | 100%
conversation-sidebar.component.ts: 87.5% | 91.66%
message-inline-input.component.ts: 75% | 93.75%
course.messages.service.ts: 71.47% | 89.28%

### Screenshots


![image](https://user-images.githubusercontent.com/67344158/161953428-19fc64c9-0ba0-4c40-bc50-02e4eb706d4d.png)

https://user-images.githubusercontent.com/67344158/161955110-0da05f07-7659-42f9-8eb2-3df5a7532170.mp4
